### PR TITLE
[CUDA graphs] Cuda RNG-safe graph capture and replay bindings

### PR DESCRIPTION
**Pull request by mcarilli**
Part 2 of https://github.com/pytorch/pytorch/pull/46148 refactor.  (part 1 was https://github.com/pytorch/pytorch/pull/48694.)
Contains
- a few more CUDAGeneratorImpl diffs to clean up graph capture interaction
- Capture and replay bindings that interact correctly with CUDAGeneratorImpl
- Tests.

Diffs compile and tests pass on my machine (ubuntu 20.04, cuda 11.0) but it needs finetuning for many CI builds.
 
See [Note [CUDA Graph-safe RNG states]](https://github.com/pytorch/pytorch/blob/02d89f9f1d7f32ebf7ec509d5c14b2f39690997a/aten/src/ATen/CUDAGeneratorImpl.h#L13-L85) for the strategy, based on https://github.com/pytorch/pytorch/pull/46148#issuecomment-724414794.

### CODE REVIEW

**Comment by ezyang** [comment]
Lotta test failures lol

File: aten/src/ATen/cuda/CUDAGraph.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+struct TORCH_CUDA_API CUDAGraph {
+  CUDAGraph();
+  ~CUDAGraph();
+
+  void capture_begin();
+  void capture_end();
+  void replay();
+
+  protected:
+  #if CUDA_VERSION >= 11000
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAGraph.h:17]: nit: no leading whitespace for preproc directives

File: aten/src/ATen/cuda/CUDAGraph.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#include <ATen/Tensor.h>
+#include <c10/core/Device.h>
+#include <c10/cuda/CUDAStream.h>
+
+namespace at {
+namespace cuda {
+
+struct TORCH_CUDA_API CUDAGraph {
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAGraph.h:8]: Probably should have a `// See Note [CUDA Graph...` reference somewhere here; probably framed as "why does PyTorch need its own CUDAGraph rep"

File: aten/src/ATen/cuda/CUDAGraph.cpp [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  // For now, a CUDAGraph instance only accommodates the default generator on the device that's
+  // current when capture begins. If any op in the captured region uses a non-default generator,
+  // or a generator on another device, the offending generator will throw an error.
+  // These restrictions simplify CUDAGraph, but could be relaxed in the future:
+  // in principle, the underlying Cuda calls do permit cross-device ops to be captured.
+  auto* gen = get_generator_or_default<CUDAGeneratorImpl>(
+      c10::nullopt, cuda::detail::getDefaultCUDAGenerator());
+
+  auto options = TensorOptions().device(at::kCUDA).dtype(at::kLong);
+  offset_extragraph_ = at::empty({1}, options);
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAGraph.cpp:34]: you sure you want empty here?

File: aten/src/ATen/cuda/CUDAGraph.cpp [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+  gen->capture_prologue(offset_extragraph_.data_ptr<int64_t>());
+
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  TORCH_CHECK(stream != at::cuda::getDefaultCUDAStream(),
+              "CUDA graphs must be captured on a non-default stream. "
+              "(However, after capture, it's ok to replay them on the "
+              "default stream.)");
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAGraph.cpp:43]: This is a CUDA graph api limitation?

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp [hunk 5]
```diff
@@ -151,25 +149,21 @@ uint64_t CUDAGeneratorImpl::philox_offset_per_thread() {
...
  * offset_intragraph tracks the offset in the graphed region.
  */
-void CUDAGeneratorImpl::graph_prologue(int64_t* offset_extragraph) {
-  TORCH_CHECK((void*)this ==
-              (void*)&at::cuda::detail::getDefaultCUDAGenerator(device_.index()),
-              CAPTURE_DEFAULT_GENS_MSG);
+void CUDAGeneratorImpl::capture_prologue(int64_t* offset_extragraph) {
   offset_extragraph_ = offset_extragraph;
   offset_intragraph_ = 0;
+  graph_expects_this_gen_ = true;
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.cpp:159]: What if you have multiple captures going on at the same time?

File: test/test_cuda.py [hunk 6]
```diff
@@ -3280,5 +3280,200 @@ class TestNamedTupleInput_1(NamedTuple):
...
             self.assertEqual(expected_b, x.b)
 
+    def test_graph_capture_simple(self):
+        s1 = torch.cuda.Stream()
+
+        with torch.cuda.stream(s1):
+            a = torch.zeros((1000,), device="cuda")
+            a += 1
+            g = torch.cuda.Graph()
+            g.capture_begin()
```
Comment (ezyang) [inline test/test_cuda.py:3290]: you sure you don't want a context manager for this?

Comment (ezyang) [reply test/test_cuda.py:3290]: The manager could also take care of getting onto a non-default stream

File: test/test_cuda.py [hunk 7]
```diff
@@ -3280,5 +3280,200 @@ class TestNamedTupleInput_1(NamedTuple):
...
+            g.capture_end()
+        torch.cuda.current_stream().wait_stream(s1)
+
+        g.replay()
+        g.replay()
+
+        self.assertTrue(a.sum().item() == 3000.)
+
+    def test_graph_rng_functional(self):
+        # The caching allocator isn't yet graph-safe.
```
Comment (ezyang) [inline test/test_cuda.py:3301]: Yeah, so what's the plan for this anyway? Thread local allocator override?

**Comment by ezyang** [comment]
The implementation bits seem fine. But the UX may need some smithing. Is there a plan on record here? (@ngimel?)

File: test/test_cuda.py (hunk 7 above)
Comment (mcarilli) [reply test/test_cuda.py:3303]: https://www.youtube.com/watch?v=Lf6aMcnR9WQ
Long term (many months) I'm working with cudaMallocAsync people to ensure cudaMallocAsync plays well with cuda graphs.  Hopefully after we integrate cudaMallocAsync, capturing allocations will "just work."

Short term, Arslan Zulfiqar, one of our people, hacked the Pytorch allocator to request and reserve a stream-silo for the graph(s):  graph capture gets its own memory silo/pool in which it can reuse memory, and it won't be affected by eager allocations.  My very next task after this PR is to look into upstreaming his changes or some variation.

File: test/test_cuda.py (hunk 6 above)
Comment (mcarilli) [reply test/test_cuda.py:3291]: I think eventually a context manager is the right exposure, but right now, this won't be documented, and exists to serve experiments.  In experiments, I don't want to exit gracefully from graphed regions if capture fails.  I want all hell to break loose so we see it.

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 4 above)
Comment (mcarilli) [reply aten/src/ATen/cuda/CUDAGraph.cpp:41]: Yes.

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 3 above)
Comment (mcarilli) [reply aten/src/ATen/cuda/CUDAGraph.cpp:32]: Yes, it doesn't matter if we run on garbage, all we're trying to do is bake in the right pointers.  If there's data dependent control flow in the graphed region, we have no business graphing it anyway.

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 4 above)
Comment (ngimel) [reply aten/src/ATen/cuda/CUDAGraph.cpp:41]: To expand on this answer a bit, it's probably because of implicit synchronization on the default stream that CUDA graph does not want to capture.

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 3 above)
Comment (mcarilli) [reply aten/src/ATen/cuda/CUDAGraph.cpp:32]: Also, deliberately beginning capture with garbage here will stress any unwanted/unexpected data dependence, and help catch failures.

**Comment by mcarilli** [comment]
> But the UX may need some smithing.  Is there a plan on record here?

Agreed, the goal of this PR is a lightweight interface that's foundation for experiments.  It won't be documented yet.  Locally, i'm experimenting with user-facing APIs to graph selected regions (right now my UX looks similar to [torch.jit.trace](https://pytorch.org/docs/stable/generated/torch.jit.trace.html)), and it works on mlperf bert, but the eventual UX is a longer conversation that requires tests on more models.

There are other balls in the air (like, right now nccl calls aren't capturable, but they will be in the future at some uncertain time) that could affect the UX.

File: aten/src/ATen/cuda/CUDAGraph.cpp [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  AT_CUDA_CHECK(cudaGraphInstantiate(&graph_exec_, graph_, NULL, NULL, 0));
+  has_graph_exec_ = true;
+
+  auto* gen = get_generator_or_default<CUDAGeneratorImpl>(
+      c10::nullopt, cuda::detail::getDefaultCUDAGenerator());
+  wholegraph_increment_ = gen->capture_epilogue();
+
+  // Now that we've instantiated graph_ into graph_exec_,
+  // we don't need graph_ anymore.
+  AT_CUDA_CHECK(cudaGraphDestroy(graph_));
```
[the remark this answers is not in the document]
Comment (mcarilli) [reply aten/src/ATen/cuda/CUDAGraph.cpp:87]: It's also handled in the CUDAGraph destructor, but I don't like that (and neither does the compiler) because it can throw an exception.

Local RAII guard here is a good idea that lets me take cudaGraphDestroy out of the destructor, but cudaGraphExecDestroy is a different story.  I don't see how to clean up graph_exec_ without cudaGraphExecDestroy in [either the destructor or a __del__ method in Python](https://github.com/pytorch/pytorch/pull/48875/files#diff-d7302d133bb5e0890fc94de9aeea4d9d442555a3b40772c9db10edb5cf36a35cR120-R134), and neither is a great option.

File: test/test_cuda.py (hunk 6 above)
Comment (mcarilli) [reply test/test_cuda.py:3291]: I'm hacking on some local UXs that take care of non-default stream.  The goal of this PR is to introduce a very lightweight capture exposure to enable UX hacking.

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp (hunk 5 above)
Comment (mcarilli) [reply aten/src/ATen/cuda/CUDAGeneratorImpl.cpp:159]: Good point.  RNG consumer calls tend to be mutexed, so I think the best approach is to make `graph_expects_this_gen_` an argument to those consumer-called functions.

**Comment by mcarilli** [comment]
> Lotta test failures lol
aye.  i pushed a change to skip tests for cuda <11, hoping that would clean some of them up...

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcarilli) [reply aten/src/ATen/cuda/CUDAGraph.cpp:87]: Uh oh, using an RAII guard for graph_ means I call `AT_CUDA_CHECK(cudaGraphDestroy(graph_));` in the guard destructor.  I can't win here :joy: :gun:

File: aten/src/ATen/cuda/CUDAGraph.h (hunk 2 above)
Comment (mcarilli) [reply aten/src/ATen/cuda/CUDAGraph.h:9]: Added [Note [CUDA Graph Wrapper Class]](https://github.com/pytorch/pytorch/pull/48875/files#diff-d7302d133bb5e0890fc94de9aeea4d9d442555a3b40772c9db10edb5cf36a35cR10-R23)

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcarilli) [reply aten/src/ATen/cuda/CUDAGraph.cpp:103]: Cuda 11_1 build refused to compile with exceptions in destructors.  Moved graph cleanup to a reset() method called by `__del__` in a thin Python wrapper (`__del__` may or may not be a lesser evil than a throwing destructor, but it's the evil CI allows).

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 3 above)
Comment (ezyang) [reply aten/src/ATen/cuda/CUDAGraph.cpp:48]: I mean, half the time the garbage is going to be zeros, if you really want to find unexpected data dependence, should fill it with some sentinel.

File: test/test_cuda.py (hunk 7 above)
Comment (ezyang) [reply test/test_cuda.py]: So, are you saying, when cudaMallocAsync becomes a thing, we get rid of PyTorch's CUDA caching allocator?

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (ezyang) [reply aten/src/ATen/cuda/CUDAGraph.cpp:103]: Yeah, just ignore the error code, we do this in a few other places too

File: test/test_cuda.py [hunk 9]
```diff
@@ -2896,6 +2896,209 @@ def test_max_large_axis(self):
...
+    @unittest.skipIf((not TEST_CUDA) or
+                     TEST_WITH_ROCM or
+                     int(torch.version.cuda.split(".")[0]) < 11, "CUDA >= 11.0 required for graphs")
+    def test_graph_capture_simple(self):
+        s1 = torch.cuda.Stream()
+
+        with torch.cuda.stream(s1):
+            a = torch.zeros((1000,), device="cuda")
+            a += 1
+            g = torch.cuda.Graph()
```
Comment (ezyang) [inline test/test_cuda.py:2908]: This API name don't look very private to me! 🤣

**Comment by ezyang** [comment]
I guess I'm OK with merging this on an experimental basis, but right now it's too public looking for my taste.

File: test/test_cuda.py (hunk 7 above)
Comment (mcarilli) [reply test/test_cuda.py]: I think that would be a sad waste, not to mention loss of control for corner cases when you need it.  I think the perfect world would be, cudaMallocAsync and the Pytorch allocator are alternative backends for some common allocator interface (which may also accept external allocators that support the same interface https://github.com/pytorch/pytorch/issues/43144)

File: test/test_cuda.py (hunk 9 above)
Comment (mcarilli) [reply test/test_cuda.py:2908]: :shushing_face: 
Alright, I can not import it as Graph for now, so usage (for experimenters) would be
```
g = torch._C._CudaGraphBase()
g.capture_begin/capture_end/replay()
```

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcarilli) [reply aten/src/ATen/cuda/CUDAGraph.cpp:103]: ? Do you mean I should go back to cleanup in the destructor, rather than `__del__`?  If so, how do I tell the CI build (I believe it was `pytorch_libtorch_linux_xenial_cuda11_1_cudnn8_py3_gcc7_build`, possibly others) not to error on the throwing destructor?  Or should I suppress the exception with a try catch?  I don't want to do the latter:  silently ignoring errors is worse than catastrophically failing.

[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/cuda/CUDAGraph.cpp:103]: Yeah, other places where we are ignoring the error happen during final cleanup when the process exits (cublas handles and the like), so who cares, whereas graph desctructor should be called during runtime more or less regularly, so I agree ignoring the code is not ideal. I vote for `__del__` unless there's a better solution.

File: test/test_cuda.py (hunk 9 above)
Comment (ngimel) [reply test/test_cuda.py:2908]: Just make it `torch.cuda._Graph` for now, rename when we have a better idea what we want to do with it.

**Comment by ezyang** [comment]
this seems ok. @ngimel do you want to land it?

**Comment by ngimel** [comment]
Let me take a last look and I will.

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/cuda/CUDAGraph.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (mcarilli) [reply aten/src/ATen/cuda/CUDAGraph.cpp:103]: per slack discussion, moved cleanup back to C++ destructor, but with warnings instead of throwing checks.
